client/libra-dev/include/data.h

```diff
@@ -0,0 +1,181 @@
+ * @param[out] caller allocated LibraSignedTransaction to write into.
+ *
+ * @returns status code, one of LibraAPIStatus
+*/
+enum LibraStatus libra_LibraSignedTransaction_from(const uint8_t *buf, size_t len, struct LibraSignedTransaction *out);
+
+/*!
+ * Get serialized raw transaction from a list of transaction parameters
+ *
+ * To get the serialized raw transaction in a memory safe manner, the client needs to pass in a pointer to a pointer to the allocated memory in rust
+ * and call free on the memory address with `libra_free_bytes_buffer`.
+ * @param[out] buf is the pointer that will be filled with the memory address of the transaction allocated in rust. User takes ownership of pointer returned by *buf, which needs to be freed using libra_free_bytes_buffer
+ * @param[out] len is the length of the raw transaction memory buffer.
+*/
+enum LibraStatus libra_RawTransactionBytes_from(const uint8_t sender[32], const uint8_t receiver[32], uint64_t sequence, uint64_t num_coins, uint64_t max_gas_amount, uint64_t gas_unit_price, uint64_t expiration_time_secs, uint8_t** buf, size_t* len);
+
+/*!
+ * This function takes in a raw transaction, public key and signature in bytes, and return a signed transaction in bytes.
+ * To get the serialized signed transaction in a memory safe manner, the client needs to pass in a pointer to a pointer to the allocated memory in rust
+ * and call free on the memory address with `libra_free_bytes_buffer`.
+ * @param[out] buf_result is the pointer that will be filled with the memory address of the transaction allocated in rust. User takes ownership of pointer returned by *buf, which needs to be freed using libra_free_bytes_buffer
+ * @param[out] len_result is the length of the signed transaction memory buffer.
+*/
+enum LibraStatus libra_RawTransaction_sign(const uint8_t *buf_raw_txn, size_t len_raw_txn, const uint8_t *buf_public_key, size_t len_public_key, const uint8_t *buf_signature, size_t len_signature, uint8_t** buf_result, size_t* len_result);
+
+/*!
+ * This function takes in private key in bytes and return the associated public key and address
+ * @param[out] out is a pointer to LibraAccountKey struct client passed in by initializing an empty LibraAccountKey struct
+*/
+enum LibraStatus libra_LibraAccount_from(const uint8_t private_key_bytes[32], struct LibraAccountKey *out);
+
+/*!
+ * This function takes in an event key, event data and event type tag in bytes, and return LibraEvent.
+ * To get the event in a memory safe manner, the client needs to call free on the output with `libra_LibraEvent_free`.
+ * @param[out] caller allocated LibraEvent to write into.
+ * @returns status code, one of LibraStatus
+*/
+enum LibraStatus libra_LibraEvent_from(const uint8_t *buf_key, size_t len_key, const uint8_t *buf_data, size_t len_data, const uint8_t *buf_type_tag, size_t len_type_tag, struct LibraEvent **out);
+void libra_LibraEvent_free(struct LibraEvent *out);
```

Comment:
Why does this `libra_*_from` have different memeory management than the others?

Reply:
The case with LibraEvent is that documentation is wrong.

We have been trying to avoid rust allocated memory model because that makes the client responsible to free it later, there are two cases where we have indetermistic amount of data to return so we have to use alloc/free model, other wise we just fill into preallocated structures.

Reply:
I think either we need to make all memory management the same (from you comments, this implies all memory is Rust allocated), or the things with different memory management semantics need to be named differently. Otherwise I think the chance of errors is too high.

If you want to do the latter, maybe Rust allocated ones are called `_new` instead of `_from`.